### PR TITLE
New ancestor/descendant algo

### LINKED ISSUE #5286: [CT-677] [Bug] Downstream models do not run correctly after NetworkX 2.8.1
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Any 1.0.x version of dbt installed after NetworkX 2.8.1 was released does not correctly run downstream models when using selection syntax.  For some reason this does not seem to effect 1.1

### Expected Behavior

All downstream models are run

### Steps To Reproduce

Install dbt-<whatever> @ v1.0.x
Create a project with a `model_foo` and downstream `model_bar`
Run dbt with selection `dbt run --select +model_foo+`
Observe that only `model_foo` is run

### Relevant log output

_No response_

### Environment

```markdown
- OS: MacOS
- Python: 3.10
- dbt: 1.0.6
```

### What database are you using dbt with?

postgres

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by iknox-fa**
resolves #5286

### Description
Updates ancestor/descendant algos to something more sane that also seems to correct an as-yet-undetermined root cause bug in either networkx or in the internal graph construction.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

### CODE REVIEW

**Review by stu-k — APPROVED** [review]
Nothing stands out, but I'd probably wait for Gerda's review before merging 👍

**Review by gshank — APPROVED** [review]
This is not really my area of expertise; it looks like it passes the tests -- so I hope we have enough tests :)

**Review by ChenyuLInx — APPROVED** [review]

**Pull request merged**
